README
		--- digup - A Digest Updating Tool ---

Description

digup is a tool to update md5sum or shasum digest files. It will read
existing digest files, check the current directory for new, updated,
modified, renamed or deleted files and query the user with a summary
of changes. After reviewing the updates, they can be written back to
the digest file.

One of the envisioned applications of digup is to update and verify
incremental archives like chronological data storages or music
collections. By using digup possibly undesired changes or lost files
can easily be detected while quickly adding new files.

Another purpose could be to automatically verify the integrity of hard
disk copies of archives, as backups to hard disks are becoming
increasingly popular. Using a full file digest scan even slowly
creeping bad blocks on old hard disks can be detected. By using a
crontab entry, this check can be performed unattended and routinely.

In normal operation only touched files with newer modification times
are fully read and their digest compared. Optionally a full scan can
be performed to test all file contents against their expected digests.

Symbolic links are supported by either following the link and reading
the target's digest or by saving only the link target path and
verifying it against the old one.

The digest files written by digup are compatible with those generated
and read by md5sum and similar programs from the coreutils
package. Additional information like file size and modification time
or symlink targets are stored on comment lines.

Four digest algorithms are supported: MD5, SHA1, SHA256 and
SHA512. The digest file itself is also checksummed using CRC32 against
unintentional changes. A fast red-black binary tree is used for the
internal file list, allowing fast operation on a large number of
files.

Website

http://idlebox.net/2009/digup/


ChangeLog
2011-01-30 19:05  Timo Bingmann

	* src/digup.c: Adding shortcut option -w, --windows for --modify-window=1 as needed to check files on FAT filesystems.
	* src/digup.c: New feature: --restrict=substring to temporarily ignore all files not matching the substring. Required a new fileinfo state FS_SKIPPED and appropriate commands.

2011-01-13 17:11  Timo Bingmann

	* src/digup.c: Fixing problems with 64-bit file sizes on win32 under (mingw32): using _stat64 on mingw now.
	* src/digup.c: Simple variable rename node -> digestiter for clarity.

2011-01-13 17:04  Timo Bingmann

	* src/digup.c: Adding check in digest_file to verify that the
	complete file was read.

2011-01-13 17:01  Timo Bingmann

	* src/digup.c: Using intptr_t in rbtree tests to fix warnings
	about pointer conversion on 64-bit systems.

2010-11-29 18:05  Timo Bingmann

	* src/digup.c: Adding new flags --modify-window inspired by rsync:
	allow modification time deltas larger than zero e.g. for backuping
	on FAT filesystems.
	* src/digup.c: Adding new option --exclude-marker=file. If file is
	found in a directory during the recursive scan, the directory
	itself and all subdirectories are skipped.

2010-10-03 11:28  Timo Bingmann

	* src/digup.c: Applying contributed patch which fixes segfault on
	amd64 due to va_start(), va_end() use pattern.

2010-08-20 10:49  Timo Bingmann

	* src/digup.c: Fixed string allocation bug causing a segfault when
	using -t followed by -f parameters.

2010-03-20 14:27  Timo Bingmann

	* src/digup.c: Changing size variable from ssize_t to long long to
	enable support for large files.

2009-11-10 14:57  Timo Bingmann

	* Initial release of 0.6.23


src/sha1.c
/*****************************************************************************
 * Functions to compute SHA1 message digest.                                 *
 *                                                                           *
 * Copyright (C) 2000, 2001, 2003, 2004, 2005, 2006, 2008                    *
 *     by Free Software Foundation, Inc.                                     *
 *                                                                           *
 * This file was taken from coreutils-7.4 and adapted for digup.             *
 *                                                                           *
 * This program is free software; you can redistribute it and/or modify it   *
 * under the terms of the GNU General Public License as published by the     *
 * Free Software Foundation; either version 3, or (at your option) any       *
 * later version.                                                            *
 *                                                                           *
 * This program is distributed in the hope that it will be useful,           *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 * GNU General Public License for more details.                              *
 *                                                                           *
 * You should have received a copy of the GNU General Public License         *
 * along with this program; if not, write to the Free Software Foundation,   *
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.        *
 *                                                                           *
 * Written by Scott G. Miller                                                *
 * Credits: Robert Klep <robert@ilse.nl>  -- Expansion function fix          *
 *****************************************************************************/

/* $Id$ */

#include "sha1.h"

#include <stddef.h>
#include <string.h>

#if HAVE_ENDIAN_H

#include <endian.h>
#define WORDS_ARE_BIGENDIAN (__BYTE_ORDER == __BIG_ENDIAN)

#elif HAVE_SYS_PARAM_H

#include <sys/param.h>
#define WORDS_ARE_BIGENDIAN (BYTE_ORDER == BIG_ENDIAN)

#else
#error "Cannot determine system endianness."
#endif

#if WORDS_ARE_BIGENDIAN
# define SWAP(n) (n)
#else
# define SWAP(n)								\
    (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#endif

#define BLOCKSIZE 4096

#if BLOCKSIZE % 64 != 0
# error "invalid BLOCKSIZE"
#endif

/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */ };


/* Take a pointer to a 160 bit block of data (five 32 bit ints) and
   initialize it to the start constants of the SHA1 algorithm.  This
   must be called before using hash in the call to sha1_hash.  */
void
sha1_init_ctx (struct sha1_ctx *ctx)
{
    ctx->A = 0x67452301;
    ctx->B = 0xefcdab89;
    ctx->C = 0x98badcfe;
    ctx->D = 0x10325476;
    ctx->E = 0xc3d2e1f0;

    ctx->total[0] = ctx->total[1] = 0;
    ctx->buflen = 0;
}

/* Copy the 4 byte value from v into the memory location pointed to by *cp,
   If your architecture allows unaligned access this is equivalent to
   * (uint32_t *) cp = v  */
static inline void
set_uint32 (char *cp, uint32_t v)
{
    memcpy (cp, &v, sizeof v);
}

/* Put result from CTX in first 20 bytes following RESBUF.  The result
   must be in little endian byte order.  */
void *
sha1_read_ctx (const struct sha1_ctx *ctx, void *resbuf)
{
    char *r = resbuf;
    set_uint32 (r + 0 * sizeof ctx->A, SWAP (ctx->A));
    set_uint32 (r + 1 * sizeof ctx->B, SWAP (ctx->B));
    set_uint32 (r + 2 * sizeof ctx->C, SWAP (ctx->C));
    set_uint32 (r + 3 * sizeof ctx->D, SWAP (ctx->D));
    set_uint32 (r + 4 * sizeof ctx->E, SWAP (ctx->E));

    return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.  */
void *
sha1_finish_ctx (struct sha1_ctx *ctx, void *resbuf)
{
    /* Take yet unprocessed bytes into account.  */
    uint32_t bytes = ctx->buflen;
    size_t size = (bytes < 56) ? 64 / 4 : 64 * 2 / 4;

    /* Now count remaining bytes.  */
    ctx->total[0] += bytes;
    if (ctx->total[0] < bytes)
	++ctx->total[1];

    /* Put the 64-bit file length in *bits* at the end of the buffer.  */
    ctx->buffer[size - 2] = SWAP ((ctx->total[1] << 3) | (ctx->total[0] >> 29));
    ctx->buffer[size - 1] = SWAP (ctx->total[0] << 3);

    memcpy (&((char *) ctx->buffer)[bytes], fillbuf, (size - 2) * 4 - bytes);

    /* Process last bytes.  */
    sha1_process_block (ctx->buffer, size * 4, ctx);

    return sha1_read_ctx (ctx, resbuf);
}

/* Compute SHA1 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
sha1_buffer (const char *buffer, size_t len, void *resblock)
{
    struct sha1_ctx ctx;

    /* Initialize the computation context.  */
    sha1_init_ctx (&ctx);

    /* Process whole buffer but last len % 64 bytes.  */
    sha1_process_bytes (buffer, len, &ctx);

    /* Put result in desired memory area.  */
    return sha1_finish_ctx (&ctx, resblock);
}

void
sha1_process_bytes (const void *buffer, size_t len, struct sha1_ctx *ctx)
{
    /* When we already have some bits in our internal buffer concatenate
       both inputs first.  */
    if (ctx->buflen != 0)
    {
	size_t left_over = ctx->buflen;
	size_t add = 128 - left_over > len ? len : 128 - left_over;

	memcpy (&((char *) ctx->buffer)[left_over], buffer, add);
	ctx->buflen += add;

	if (ctx->buflen > 64)
	{
	    sha1_process_block (ctx->buffer, ctx->buflen & ~63, ctx);

	    ctx->buflen &= 63;
	    /* The regions in the following copy operation cannot overlap.  */
	    memcpy (ctx->buffer,
		    &((char *) ctx->buffer)[(left_over + add) & ~63],
		    ctx->buflen);
	}

	buffer = (const char *) buffer + add;
	len -= add;
    }

    /* Process available complete blocks.  */
    if (len >= 64)
    {
#if !_STRING_ARCH_unaligned
# define alignof(type) offsetof (struct { char c; type x; }, x)
# define UNALIGNED_P(p) (((size_t) p) % alignof (uint32_t) != 0)
	if (UNALIGNED_P (buffer))
	    while (len > 64)
	    {
		sha1_process_block (memcpy (ctx->buffer, buffer, 64), 64, ctx);
		buffer = (const char *) buffer + 64;
		len -= 64;
	    }
	else
#endif
	{
	    sha1_process_block (buffer, len & ~63, ctx);
	    buffer = (const char *) buffer + (len & ~63);
	    len &= 63;
	}
    }

    /* Move remaining bytes in internal buffer.  */
    if (len > 0)
    {
	size_t left_over = ctx->buflen;

	memcpy (&((char *) ctx->buffer)[left_over], buffer, len);
	left_over += len;
	if (left_over >= 64)
	{
	    sha1_process_block (ctx->buffer, 64, ctx);
	    left_over -= 64;
	    memcpy (ctx->buffer, &ctx->buffer[16], left_over);
	}
	ctx->buflen = left_over;
    }
}

/* --- Code below is the primary difference between md5.c and sha1.c --- */

/* SHA1 round constants */
#define K1 0x5a827999
#define K2 0x6ed9eba1
#define K3 0x8f1bbcdc
#define K4 0xca62c1d6

/* Round functions.  Note that F2 is the same as F4.  */
#define F1(B,C,D) ( D ^ ( B & ( C ^ D ) ) )
#define F2(B,C,D) (B ^ C ^ D)
#define F3(B,C,D) ( ( B & C ) | ( D & ( B | C ) ) )
#define F4(B,C,D) (B ^ C ^ D)

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.
   Most of this code comes from GnuPG's cipher/sha1.c.  */

void
sha1_process_block (const void *buffer, size_t len, struct sha1_ctx *ctx)
{
    const uint32_t *words = buffer;
    size_t nwords = len / sizeof (uint32_t);
    const uint32_t *endp = words + nwords;
    uint32_t x[16];
    uint32_t a = ctx->A;
    uint32_t b = ctx->B;
    uint32_t c = ctx->C;
    uint32_t d = ctx->D;
    uint32_t e = ctx->E;

    /* First increment the byte count.  RFC 1321 specifies the possible
       length of the file up to 2^64 bits.  Here we only compute the
       number of bytes.  Do a double word increment.  */
    ctx->total[0] += len;
    if (ctx->total[0] < len)
	++ctx->total[1];

#define rol(x, n) (((x) << (n)) | ((uint32_t) (x) >> (32 - (n))))

#define M(I) ( tm =   x[I&0x0f] ^ x[(I-14)&0x0f] \
		    ^ x[(I-8)&0x0f] ^ x[(I-3)&0x0f] \
	       , (x[I&0x0f] = rol(tm, 1)) )

#define R(A,B,C,D,E,F,K,M)  do { E += rol( A, 5 )     \
				      + F( B, C, D )  \
				      + K	      \
				      + M;	      \
				 B = rol( B, 30 );    \
			       } while(0)

    while (words < endp)
    {
	uint32_t tm;
	int t;
	for (t = 0; t < 16; t++)
	{
	    x[t] = SWAP (*words);
	    words++;
	}

	R( a, b, c, d, e, F1, K1, x[ 0] );
	R( e, a, b, c, d, F1, K1, x[ 1] );
	R( d, e, a, b, c, F1, K1, x[ 2] );
	R( c, d, e, a, b, F1, K1, x[ 3] );
	R( b, c, d, e, a, F1, K1, x[ 4] );
	R( a, b, c, d, e, F1, K1, x[ 5] );
	R( e, a, b, c, d, F1, K1, x[ 6] );
	R( d, e, a, b, c, F1, K1, x[ 7] );
	R( c, d, e, a, b, F1, K1, x[ 8] );
	R( b, c, d, e, a, F1, K1, x[ 9] );
	R( a, b, c, d, e, F1, K1, x[10] );
	R( e, a, b, c, d, F1, K1, x[11] );
	R( d, e, a, b, c, F1, K1, x[12] );
	R( c, d, e, a, b, F1, K1, x[13] );
	R( b, c, d, e, a, F1, K1, x[14] );
	R( a, b, c, d, e, F1, K1, x[15] );
	R( e, a, b, c, d, F1, K1, M(16) );
	R( d, e, a, b, c, F1, K1, M(17) );
	R( c, d, e, a, b, F1, K1, M(18) );
	R( b, c, d, e, a, F1, K1, M(19) );
	R( a, b, c, d, e, F2, K2, M(20) );
	R( e, a, b, c, d, F2, K2, M(21) );
	R( d, e, a, b, c, F2, K2, M(22) );
	R( c, d, e, a, b, F2, K2, M(23) );
	R( b, c, d, e, a, F2, K2, M(24) );
	R( a, b, c, d, e, F2, K2, M(25) );
	R( e, a, b, c, d, F2, K2, M(26) );
	R( d, e, a, b, c, F2, K2, M(27) );
	R( c, d, e, a, b, F2, K2, M(28) );
	R( b, c, d, e, a, F2, K2, M(29) );
	R( a, b, c, d, e, F2, K2, M(30) );
	R( e, a, b, c, d, F2, K2, M(31) );
	R( d, e, a, b, c, F2, K2, M(32) );
	R( c, d, e, a, b, F2, K2, M(33) );
	R( b, c, d, e, a, F2, K2, M(34) );
	R( a, b, c, d, e, F2, K2, M(35) );
	R( e, a, b, c, d, F2, K2, M(36) );
	R( d, e, a, b, c, F2, K2, M(37) );
	R( c, d, e, a, b, F2, K2, M(38) );
	R( b, c, d, e, a, F2, K2, M(39) );
	R( a, b, c, d, e, F3, K3, M(40) );
	R( e, a, b, c, d, F3, K3, M(41) );
	R( d, e, a, b, c, F3, K3, M(42) );
	R( c, d, e, a, b, F3, K3, M(43) );
	R( b, c, d, e, a, F3, K3, M(44) );
	R( a, b, c, d, e, F3, K3, M(45) );
	R( e, a, b, c, d, F3, K3, M(46) );
	R( d, e, a, b, c, F3, K3, M(47) );
	R( c, d, e, a, b, F3, K3, M(48) );
	R( b, c, d, e, a, F3, K3, M(49) );
	R( a, b, c, d, e, F3, K3, M(50) );
	R( e, a, b, c, d, F3, K3, M(51) );
	R( d, e, a, b, c, F3, K3, M(52) );
	R( c, d, e, a, b, F3, K3, M(53) );
	R( b, c, d, e, a, F3, K3, M(54) );
	R( a, b, c, d, e, F3, K3, M(55) );
	R( e, a, b, c, d, F3, K3, M(56) );
	R( d, e, a, b, c, F3, K3, M(57) );
	R( c, d, e, a, b, F3, K3, M(58) );
	R( b, c, d, e, a, F3, K3, M(59) );
	R( a, b, c, d, e, F4, K4, M(60) );
	R( e, a, b, c, d, F4, K4, M(61) );
	R( d, e, a, b, c, F4, K4, M(62) );
	R( c, d, e, a, b, F4, K4, M(63) );
	R( b, c, d, e, a, F4, K4, M(64) );
	R( a, b, c, d, e, F4, K4, M(65) );
	R( e, a, b, c, d, F4, K4, M(66) );
	R( d, e, a, b, c, F4, K4, M(67) );
	R( c, d, e, a, b, F4, K4, M(68) );
	R( b, c, d, e, a, F4, K4, M(69) );
	R( a, b, c, d, e, F4, K4, M(70) );
	R( e, a, b, c, d, F4, K4, M(71) );
	R( d, e, a, b, c, F4, K4, M(72) );
	R( c, d, e, a, b, F4, K4, M(73) );
	R( b, c, d, e, a, F4, K4, M(74) );
	R( a, b, c, d, e, F4, K4, M(75) );
	R( e, a, b, c, d, F4, K4, M(76) );
	R( d, e, a, b, c, F4, K4, M(77) );
	R( c, d, e, a, b, F4, K4, M(78) );
	R( b, c, d, e, a, F4, K4, M(79) );

	a = ctx->A += a;
	b = ctx->B += b;
	c = ctx->C += c;
	d = ctx->D += d;
	e = ctx->E += e;
    }
}

/*****************************************************************************/
